hellsemble/hellsemble.py

```diff
@@ -405,34 +536,99 @@ def __fit_estimators_greedy(
                     best_model, X_val
                 )
                 failed_observations_mask_val = val_predictions != y_val
+
                 # Create prediction history entry
-                failed_observations_idx_fit = failed_observations_idx_fit[
-                    failed_observations_mask_fit
-                ]
+                failed_observations_idx_fit = fit_indices[failed_observations_mask_fit]
                 train_fitting_history_entry = np.full((X_train.shape[0]), False)
-
                 train_fitting_history_entry[failed_observations_idx_fit] = True
                 train_fitting_history.append(train_fitting_history_entry)
 
-                failed_observations_idx_val = failed_observations_idx_val[
-                    failed_observations_mask_val
-                ]
+                failed_observations_idx_val = val_indices[failed_observations_mask_val]
                 validation_fitting_history_entry = np.full(
                     (X_validation.shape[0]), False
                 )
                 validation_fitting_history_entry[failed_observations_idx_val] = True
                 validation_fitting_history.append(validation_fitting_history_entry)
 
+                # --- Regularization: allow some good observations to pass through (train) ---
+                train_score = self.evaluate_hellsemble(X_train, y_train)
```

Comment:
this part of code is 1:1 with greedy and sequential I guess, maybe it should be placed somewhere else in one function